File: sdk/typescript/src/controls.ts
import fs from "node:fs";
import path from "node:path";
import yaml from "js-yaml";
import type { Control } from "./types.js";

function walk(dir: string, out: string[] = []): string[] {
  if (!fs.existsSync(dir)) return out;
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) walk(full, out);
    else if (entry.isFile() && full.endsWith(".yaml")) out.push(full);
  }
  return out;
}

export class Controls {
  private items: Control[] = [];

  constructor(private repoRoot: string) {}

  load(): this {
    const root = path.join(this.repoRoot, "domains");
    for (const p of walk(root)) {
      try {
        const doc = yaml.load(fs.readFileSync(p, "utf8")) as Control | null;
        if (doc && doc.kind === "Control") this.items.push(doc);
      } catch {
        // skip malformed; conformance CLI is the canonical validator
      }
    }
    return this;
  }

  all(): Control[] {
    return [...this.items];
  }

  byId(id: string): Control | undefined {
    return this.items.find((c) => c.metadata.id === id);
  }

  byUcid(ucid: string): Control[] {
    return this.items.filter((c) => c.metadata.ucid === ucid);
  }

  byStatus(status: Control["metadata"]["status"]): Control[] {
    return this.items.filter((c) => c.metadata.status === status);
  }

  byDomain(domain: string): Control[] {
    return this.items.filter((c) =>
      c.metadata.id.startsWith(domainPrefix(domain))
    );
  }
}

function domainPrefix(domain: string): string {
  const map: Record<string, string> = {
    "data-governance": "DG-",
    "human-oversight": "HO-",
    "logging-traceability": "LOG-",
    "model-lifecycle": "ML-",
    "security-robustness": "SR-",
    "incident-response": "IR-",
    "transparency-disclosure": "TD-",
    "risk-management-system": "RMS-",
    "third-party-and-supply-chain": "TPS-",
    "post-market-monitoring": "PMM-",
  };
  return map[domain] ?? "";
}

File: sdk/python/src/umbrella_sdk/controls.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

_DOMAIN_PREFIX = {
    "data-governance": "DG-",
    "human-oversight": "HO-",
    "logging-traceability": "LOG-",
    "model-lifecycle": "ML-",
    "security-robustness": "SR-",
    "incident-response": "IR-",
    "transparency-disclosure": "TD-",
    "risk-management-system": "RMS-",
    "third-party-and-supply-chain": "TPS-",
    "post-market-monitoring": "PMM-",
}


@dataclass
class Control:
    data: dict[str, Any]

    @property
    def id(self) -> str:
        return self.data.get("metadata", {}).get("id", "")

    @property
    def ucid(self) -> str:
        return self.data.get("metadata", {}).get("ucid", "")

    @property
    def status(self) -> str:
        return self.data.get("metadata", {}).get("status", "")

    @property
    def name(self) -> str:
        return self.data.get("metadata", {}).get("name", "")


class Controls:
    def __init__(self, repo_root: Path) -> None:
        self._root = repo_root
        self._items: list[Control] = []

    def load(self) -> Controls:
        domains = self._root / "domains"
        if not domains.exists():
            return self
        for p in domains.rglob("*.yaml"):
            try:
                doc = yaml.safe_load(p.read_text())
            except yaml.YAMLError:
                continue
            if isinstance(doc, dict) and doc.get("kind") == "Control":
                self._items.append(Control(doc))
        return self

    def all(self) -> list[Control]:
        return list(self._items)

    def by_id(self, control_id: str) -> Control | None:
        for c in self._items:
            if c.id == control_id:
                return c
        return None

    def by_ucid(self, ucid: str) -> list[Control]:
        return [c for c in self._items if c.ucid == ucid]

    def by_status(self, status: str) -> list[Control]:
        return [c for c in self._items if c.status == status]

    def by_domain(self, domain: str) -> list[Control]:
        prefix = _DOMAIN_PREFIX.get(domain, "")
        if not prefix:
            return []
        return [c for c in self._items if c.id.startswith(prefix)]
